pkg/common/consumer/consumer_manager.go

```diff
@@ -0,0 +1,391 @@
+	CloseConsumerGroup(groupId string) error
+	Errors(groupId string) <-chan error
+	IsManaged(groupId string) bool
+}
+
+// groupMap is a mapping of GroupIDs to managed Consumer Group pointers (fields in the managedGroup
+// are modified directly by the kafkaConsumerGroupManagerImpl receiver functions; the pointer makes
+// removing and re-adding them to the map unnecessary in those cases)
+type groupMap map[string]*managedGroup
+
+// kafkaConsumerGroupManagerImpl is the primary implementation of a KafkaConsumerGroupManager, which
+// handles control protocol messages and stopping/starting ("pausing/resuming") of ConsumerGroups.
+type kafkaConsumerGroupManagerImpl struct {
+	logger    *zap.Logger
+	server    controlprotocol.ServerHandler
+	factory   *kafkaConsumerGroupFactoryImpl
+	groups    groupMap
+	groupLock sync.RWMutex // Synchronizes write access to the groupMap
+}
+
+// Verify that the kafkaConsumerGroupManagerImpl satisfies the KafkaConsumerGroupManager interface
+var _ KafkaConsumerGroupManager = (*kafkaConsumerGroupManagerImpl)(nil)
+
+// NewConsumerGroupManager returns a new kafkaConsumerGroupManagerImpl as a KafkaConsumerGroupManager interface
+func NewConsumerGroupManager(logger *zap.Logger, serverHandler controlprotocol.ServerHandler, brokers []string, config *sarama.Config) KafkaConsumerGroupManager {
+
+	manager := &kafkaConsumerGroupManagerImpl{
+		logger:    logger,
+		server:    serverHandler,
+		groups:    make(groupMap),
+		factory:   &kafkaConsumerGroupFactoryImpl{addrs: brokers, config: config},
+		groupLock: sync.RWMutex{},
+	}
+
+	logger.Info("Registering Consumer Group Manager Control-Protocol Handlers")
+
+	// Add a handler that understands the StopConsumerGroupOpCode and stops the requested group
+	serverHandler.AddAsyncHandler(
+		commands.StopConsumerGroupOpCode,
+		commands.StopConsumerGroupResultOpCode,
+		&commands.ConsumerGroupAsyncCommand{},
+		func(ctx context.Context, commandMessage ctrlservice.AsyncCommandMessage) {
+			processAsyncGroupNotification(commandMessage, manager.stopConsumerGroup)
+		})
+
+	// Add a handler that understands the StartConsumerGroupOpCode and starts the requested group
+	serverHandler.AddAsyncHandler(
+		commands.StartConsumerGroupOpCode,
+		commands.StartConsumerGroupResultOpCode,
+		&commands.ConsumerGroupAsyncCommand{},
+		func(ctx context.Context, commandMessage ctrlservice.AsyncCommandMessage) {
+			processAsyncGroupNotification(commandMessage, manager.startConsumerGroup)
+		})
+
+	return manager
+}
+
+// Reconfigure will incorporate a new set of brokers and Sarama config settings into the manager
+// without requiring a new control-protocol server or losing the current map of managed groups.
+// It will stop and start all of the managed groups in
+func (m *kafkaConsumerGroupManagerImpl) Reconfigure(brokers []string, config *sarama.Config) error {
```

Comment:
in what?  the suspense is killing me!  ; )

Reply:
addressed prepositional failure